### PR TITLE
controller factory: allow conditional queue key add

### PR DESCRIPTION
**Pull request by stlaz**
The conditional work queue pushing simplifies the logic for
controllers that take a certain resource type but only want to
react to the resource changes when a condition is met,
e.g. an annotation is set to a certain value.

Would be useful at https://github.com/openshift/service-ca-operator/pull/121/files#diff-eec55cf875708a3ee2197838c800c3faR80

### CODE REVIEW

File: pkg/controller/factory/factory.go [hunk 1]
```diff
@@ -97,6 +101,13 @@ func (f *Factory) WithBareInformers(informers ...Informer) *Factory {
 // is called.
 // Pass the queueKeyFn you want to use to transform the informer runtime.Object into string key used by work queue.
 func (f *Factory) WithInformersQueueKeyFunc(queueKeyFn ObjectQueueKeyFunc, informers ...Informer) *Factory {
+	return f.WithInformersCondQueueKeyFunc(
+		func(obj runtime.Object) (string, bool) { return queueKeyFn(obj), true },
+		informers...,
+	)
+}
+
+func (f *Factory) WithInformersCondQueueKeyFunc(queueKeyFn ConditionalObjectQueueKeyFunc, informers ...Informer) *Factory {
```
Comment (mfojtik) [inline pkg/controller/factory/factory.go:110]: make it ugly! WithInformersConditionalQueueKeyFunc``

Comment (mfojtik) [inline pkg/controller/factory/factory.go:110]: please add godoc and include example

File: pkg/controller/factory/factory.go [hunk 2]
```diff
@@ -82,6 +92,18 @@ func (f *Factory) WithInformers(informers ...Informer) *Factory {
 	return f
 }
 
+// WithFilteredEventsInformers is used to register event handlers and get the caches synchronized functions.
+// Pass the informers you want to use to react to changes on resources. If informer event is observed, then the Sync() function
+// is called.
+// Pass filter to filter out events that should not trigger Sync() call.
+func (f *Factory) WithFilteredEventsInformers(filter EventFilterFunc, informers ...Informer) *Factory {
+	f.filteredInformers = append(f.filteredInformers, filteredInformers{
```
Comment (sttts) [inline pkg/controller/factory/factory.go:100]: having two informer lists is looking pretty strange. Can't we unify that?

Comment (stlaz) [reply pkg/controller/factory/factory.go:100]: sure can, I'll merge those

**Comment by stlaz** [comment]
fixed an issue where the filters were not in fact passed to the `eventHandler()`

File: pkg/controller/factory/controller_context.go [hunk 3]
```diff
@@ -44,62 +44,74 @@ func (c syncContext) Recorder() events.Recorder {
...
+// as a filter for objects that are namespaces and they concern us in any way
+type namespaceChecker struct {
+	interestingNamespaces sets.String
+}
+
+func namespaceCheckingFunc(interestingNamespaces []string) func(obj interface{}) bool {
+	return namespaceChecker{
+		interestingNamespaces: sets.NewString(interestingNamespaces...),
+	}.isInterestingNamespace
+}
```
Comment (sttts) [inline pkg/controller/factory/controller_context.go:98]: use a closure directly.

```
func namespaceCheckingFunc(interestingNamespaces []string) func(obj interface{}) bool {
  return func(obj interface{}) bool {
    ...
  }
}
```

File: pkg/controller/factory/factory.go [hunk 4]
```diff
@@ -78,7 +87,19 @@ func (f *Factory) WithSync(syncFn SyncFunc) *Factory {
 // Pass informers you want to use to react to changes on resources. If informer event is observed, then the Sync() function
 // is called.
 func (f *Factory) WithInformers(informers ...Informer) *Factory {
-	f.informers = append(f.informers, informers...)
+	f.informers = append(f.informers, filteredInformers{informers: informers})
```
Comment (sttts) [inline pkg/controller/factory/factory.go:90]: call WithFilteredEventsInformers here with nil

File: pkg/controller/factory/controller_context.go (hunk 3 above)
Comment (stlaz) [reply pkg/controller/factory/controller_context.go:98]: it's now returning closure :+1:

**Pull request merged**